server/src/main/java/org/elasticsearch/cluster/coordination/LeaderChecker.java

```diff
@@ -0,0 +1,268 @@
+import org.elasticsearch.common.settings.Setting;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.unit.TimeValue;
+import org.elasticsearch.tasks.Task;
+import org.elasticsearch.threadpool.ThreadPool.Names;
+import org.elasticsearch.transport.TransportChannel;
+import org.elasticsearch.transport.TransportException;
+import org.elasticsearch.transport.TransportRequest;
+import org.elasticsearch.transport.TransportRequestOptions;
+import org.elasticsearch.transport.TransportRequestOptions.Type;
+import org.elasticsearch.transport.TransportResponse;
+import org.elasticsearch.transport.TransportResponse.Empty;
+import org.elasticsearch.transport.TransportResponseHandler;
+import org.elasticsearch.transport.TransportService;
+
+import java.io.IOException;
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
+
+public class LeaderChecker extends AbstractComponent {
+
+    public static final String LEADER_CHECK_ACTION_NAME = "internal:coordination/fault_detection/leader_check";
+
+    // the time between checks sent to the leader
+    public static final Setting<TimeValue> LEADER_CHECK_INTERVAL_SETTING =
+        Setting.timeSetting("cluster.fault_detection.leader_check.interval",
+            TimeValue.timeValueMillis(10000), TimeValue.timeValueMillis(1000), Setting.Property.NodeScope);
+
+    // the timeout for each check sent to the leader
+    public static final Setting<TimeValue> LEADER_CHECK_TIMEOUT_SETTING =
+        Setting.timeSetting("cluster.fault_detection.leader_check.timeout",
+            TimeValue.timeValueMillis(10000), TimeValue.timeValueMillis(1), Setting.Property.NodeScope);
+
+    // the number of failed checks that must happen before the leader is considered to have failed.
+    public static final Setting<Integer> LEADER_CHECK_RETRY_COUNT_SETTING =
+        Setting.intSetting("cluster.fault_detection.leader_check.retry_count", 3, 1, Setting.Property.NodeScope);
+
+    private final TimeValue leaderCheckInterval;
+    private final TimeValue leaderCheckTimeout;
+    private final int leaderCheckRetryCount;
+    private final TransportService transportService;
+    private final Runnable onLeaderFailure;
+
+    private volatile DiscoveryNodes lastPublishedDiscoveryNodes;
+
+    public LeaderChecker(final Settings settings, final TransportService transportService, final Runnable onLeaderFailure) {
+        super(settings);
+        leaderCheckInterval = LEADER_CHECK_INTERVAL_SETTING.get(settings);
+        leaderCheckTimeout = LEADER_CHECK_TIMEOUT_SETTING.get(settings);
+        leaderCheckRetryCount = LEADER_CHECK_RETRY_COUNT_SETTING.get(settings);
+        this.transportService = transportService;
+        this.onLeaderFailure = onLeaderFailure;
+
+        transportService.registerRequestHandler(LEADER_CHECK_ACTION_NAME, Names.GENERIC, LeaderCheckRequest::new, this::handleLeaderCheck);
```

Comment:
`handleLeaderCheck` is so light-weight it could run on the network thread

Reply:
Sure, I pushed c62ec25f47ce13af862f212dfbb1470e18dde27c